File: AlphaGo/AlphaGo/MCTS.py
import numpy as np
import random
import math


class MCTS:
    def __init__(self):
        self.states = {}  # visits, [actions]
        self.state_action = {}  # [visits, sum q values]

    def get_recommended_move(self, state):
        max_visits = None
        best_action = None
        for action in self.states[state][1]:
            visits = self.state_action[state + '_' + action][0]
            if max_visits == None or visits > max_visits:
                max_visits = visits
                best_action = action
        return best_action

    def get_search_statistics(self, state):
        return self.states[state], [self.state_action[state + '_' + action] for action in self.states[state][1]]

    def set_exp_param(self, exp_param):
        self.exp_param = exp_param

    def search(self, game, search_num):
        game.store_state()
        for _ in range(search_num):
            self._single_search(game)
            game.load_state()

    def _single_search(self, game):
        if game.is_finished():
            return game.outcome()

        state = game.get_state()[0]

        # Tree search
        if state in self.states:
            outcome = self._tree_search(game, state)
            return outcome

        # Node expansion
        self._node_expansion(game, state)

        # Leaf evaluation
        action, evaluation = self.rollout(game)

        # Backpropagation
        self.states[state][0] = self.states[state][0] + 1
        self.state_action[state + '_' + action] = [1, evaluation[game.get_turn()]]
        return evaluation

    def _tree_search(self, game, state):
        # _select_move employs the tree policy (uct) when choosing action
        move = self._select_move(game, state)
        game.execute_move(move)
        outcome = self._single_search(game)
        game.undo_move()

        self.states[state][0] = self.states[state][0] + 1
        move_statistics = self.state_action[state + '_' + move]
        move_statistics[0] += 1
        move_statistics[1] += outcome[game.get_turn()]
        self.state_action[state + '_' + move] = move_statistics
        return outcome

    def _select_move(self, game, state):
        max_val = None
        max_action = None

        for action in game.get_legal_moves():
            print(state)
            score = MCTS.uct(self.exp_param, self.states[state], self.state_action[state + '_' + action])
            if score == None: return action
            if max_val == None or score > max_val:
                max_val = score
                max_action = action
        return max_action

    def _node_expansion(self, game, state):
        self.states[state] = [0, []]
        moves = game.get_legal_moves()
        self.states[state][1] = moves
        for action in moves:
            self.state_action[state + '_' + action] = [0, 0]

    def rollout(self, game):
        # A random behaviour policy
        first_action = None
        move_count = 0
        while not game.is_finished():
            moves = game.get_legal_moves()
            rand = random.randint(0, len(moves) - 1)
            if move_count == 0: first_action = moves[rand]
            game.execute_move(moves[rand])
            move_count += 1

        for _ in range(move_count):
            game.undo_move()
        return first_action, game.outcome()

    @staticmethod
    def uct(c, parent, child):

        if child[0] == 0: return None
        exploration = c * math.sqrt(math.log(parent[0]) / child[0])
        q_value = child[1] / child[0]
        return q_value + exploration

    def reset(self):
        self.states = {}
        self.state_action = {}


if __name__ == '__main__':
    pass

File: AlphaGo/Games/MCTS.cpp
#include "Hex.h"
#include <unordered_map>
#include <chrono>
#include <cstdlib>
#include <pthread.h>
#include <thread>

#define NEGINF -100000


class MCTS {
	private:
		
		unordered_map<string, int> stateVisits; // key = state : value = times visited
		unordered_map<string, vector<string>> actions; // key = state : values = actions
		unordered_map<string, vector<float>> stateAction; // key = state_action : values = {visits, sum q values}
		float expParam;
		Hex game;

		pair<int, int> singleSearch() {
			//cout << game.isFinished() << endl;
			//cout << "test" << endl;

			if (game.isFinished()) {
				//cout << "finished " << game.outcome().first << endl;
				//game.printBoard();
				return game.outcome();
			}

			string state = game.getState();

			// Tree search
			if (stateVisits.find(state) != stateVisits.end())
				return treeSearch(state);
			
			
			// Node expansion
			nodeExpansion(state);

			//Leaf evaluation
			pair<string, pair<float, float>> evaluation = rollout();			
			string action = evaluation.first;
			
			float outcome = (game.getTurn() == 1) ? evaluation.second.first : evaluation.second.second;
			//if (outcome > 0)
			//	cout << endl;
			
			//Backpropagation
			stateVisits[state] += 1;
			stateAction[state + "_" + action] = vector<float> {1, outcome};
			return evaluation.second;
		}

		pair<int, int> treeSearch(string state) {
			string move = selectAction(state);
			//cout << "treeSearch " << state << endl;
			game.executeMove(move);
			pair<int, int> outcome = singleSearch();
			game.undoMove();
			int playerOutcome = (game.getTurn() == 1) ? outcome.first : outcome.second;

			stateVisits[state] += 1;
			string stateActionName = state + "_" + move;
			stateAction[stateActionName][0] += 1;
			stateAction[stateActionName][1] += playerOutcome;
			return outcome;
		}

		string selectAction(string state) {
			float maxVal = NEGINF;
			string maxAction = "";
			//cout << state << " test" << endl;
			vector<string>* legalMoves = game.getLegalMoves(new vector<string>());

			for (auto action : (*legalMoves)) {
				float score = MCTS::uct(expParam, stateVisits[state], stateAction[state + "_" + action]);
				if (score == NEGINF) 
					return action;
				if (maxVal == NEGINF || score > maxVal) {
					maxVal = score;
					maxAction = action;
				}
			}
			return maxAction;
		}

		void nodeExpansion(string state) {
			stateVisits[state] = 0;
			vector<string>* legalActions = game.getLegalMoves(new vector<string>());
			actions[state] = (*legalActions);
			
			for (auto action : (*legalActions)) {
				string stateActionName = state + "_" + action;
				stateAction[stateActionName] = vector<float> {0, 0};
			}
		}

		pair<string, pair<float, float>> rollout() {
			string firstAction = "";
			int moveCount = 0;

			while (!game.isFinished()) {
				//game.printBoard();
				//cout << "getting moves "<< endl;
				vector<string>* moves = game.getLegalMoves(new vector<string>());
				//cout << "got moves " << moves.size()<< endl;
				int actionNum = rand() % moves->size();
				if (moveCount == 0)
					firstAction = (*moves)[actionNum];
				//cout << "rollout " << game.getState()<< endl;
				game.executeMove((*moves)[actionNum]);
				moveCount++;
				//cout << "executed"<< endl;
			}
			pair<float, float> outcome = game.outcome();
			//cout << "finish rollout " << game.getState()<< endl;

			for (int x = 0; x < moveCount; x++) {
				game.undoMove();
			}
			//cout << moveCount << " " << outcome.first << " " << game.getTurn() << endl;
			return make_pair(firstAction, outcome);
		}

		static float uct(float c, int parentVisits, vector<float> child) {
			//cout << child[0] << " " << child[1] << " " << parentVisits << endl;
			if (child[0] == 0)
				return NEGINF;
			float exploration = c * sqrt(log(parentVisits) / child[0]);
			float qValue = child[1] / child[0];
			return qValue + exploration;
		}

	public:
		MCTS() : game(5,1) {
			cout << "Starting" << endl;
		}

		string getRecommendedMove(string state){
			int maxVisits = -1;
			string bestAction = "";

			for (auto action : actions[state]) {
				int visits = stateAction[state + "_" + action][0];
				
				if (maxVisits == -1 || visits > maxVisits) {
					maxVisits = visits;
					bestAction = action;
				}
			}
			return bestAction;
		}

		pair<int, vector<vector<float>>> getSearchStatistics(string state) {
			vector<vector<float>> statistics;
			cout << stateVisits[state] << endl;
			for (auto action : actions[state]) {
				statistics.push_back(stateAction[state + "_" + action]);
				cout << statistics.back()[0] << " " << statistics.back()[1] << endl;
			}
			return make_pair(stateVisits[state], statistics);
		}

		void setExpParam(float inpExpParam) {
			expParam = inpExpParam;
		}

		void setGame(Hex inpGame){
			game = inpGame;
		}

		void search(int searchNum) {
			//TODO: store and load should not be needed
			//game.storeState();
			
			for (int x = 0; x < searchNum; x++) {
				singleSearch();
				//game.loadState();
			}
		}
};

void multithread(){
	MCTS treeSearch;
	Hex game(4, 1);
	Hexagonal game2OfHex(4);
	game.setHexagonal(&game2OfHex);
	treeSearch.setExpParam(1.0);
	treeSearch.setGame(game);
	treeSearch.search(100000);
	treeSearch.getSearchStatistics("");

}

int main(){
	// TODO: 
	// fix some bug somewhere:(
	// speed up code: use pointers for vectors
	int size = 1;
	thread threads[size];
	auto start = chrono::high_resolution_clock::now();
	for (int x = 0; x < size; x++){
		threads[x] = thread(multithread);
	}
	for (int x = 0; x < size; x++){
		threads[x].join();
	}
	auto stop = chrono::high_resolution_clock::now();
	cout << chrono::duration_cast<chrono::microseconds>(stop - start).count() << endl;

	system("pause");
}
